src/iterators/chunked_read_ahead_iterator.rs

```diff
@@ -0,0 +1,442 @@
+use std::any::Any;
+use std::panic::{catch_unwind, resume_unwind, AssertUnwindSafe};
+use std::sync::mpsc::{sync_channel, Receiver};
+use std::thread::{self, JoinHandle};
+use std::vec::IntoIter;
+
+// type aliased to get clippy to not think this is too complex
+type PanicUnwindErr = Box<dyn Any + Send>;
+
+/// Iterator extension that spawns an additional thread to read-ahead in the iterator. Sends
+/// results back to this object via a channel and returns them in the same manner as a normal
+/// iterator. This is useful in the context in which the reading of an iterator (or iterators) is
+/// time consuming (e.g. reading from a compressed FASTQ file) and the main thread is bottlenecked
+/// by the speed of the underlying iterator.
+///
+/// To use on a struct that implements ``IntoIter``, it is as simple as:
+/// ```
+/// use fgoxide::iterators::chunked_read_ahead_iterator::IntoChunkedReadAheadIterator;
+///
+/// let v = vec![0,1,2,3,4,5,6,7];
+/// let chunk_size = 5;
+/// let buffer_size = 5;
+///
+/// let mut chunked_iter = v.into_iter().read_ahead(chunk_size, buffer_size);
+/// assert_eq!(chunked_iter.next(), Some(0));
+/// assert_eq!(chunked_iter.next(), Some(1));
+/// assert_eq!(chunked_iter.next(), Some(2));
+/// assert_eq!(chunked_iter.next(), Some(3));
+/// assert_eq!(chunked_iter.next(), Some(4));
+/// assert_eq!(chunked_iter.next(), Some(5));
+/// assert_eq!(chunked_iter.next(), Some(6));
+/// assert_eq!(chunked_iter.next(), Some(7));
+/// assert_eq!(chunked_iter.next(), None);
+/// ```
+/// Where ``chunk_size`` is the number of elements in the iter to include per send / recieve over
+/// the underlying channel, and ``buffer_size`` is the maximum number of chunks to keep on the
```

Comment:
I'd still prefer to call `buffer_size` something more like `chunk_count` or `chunk_queue_size`, because I think `buffer_size` is more easily misinterpreted if one isn't paying attention.